core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -349,87 +355,40 @@ public Context(
             this.timeoutInMilliseconds = timeoutInMilliseconds;
             this.session = session;
             this.warningCollector = warningCollector;
+            this.iterativeOptimizerStatsCollector = new PlanOptimizersStatsCollector(5);
```

Comment:
We shouldn't be using 5 limit here anymore as we want to merge everything back to the query level collector.
You can do the top 5 part in checkTimeoutNotExhausted, the limit could be a parameter to `getQueryRulesStatsSnapshot` as well.